### PR TITLE
Add separator to AppendPattern assembly rule

### PR DESCRIPTION
**Pull request by yakivy**
PR adds separator to AppendPattern assembly rule.

Implementation and motivation are identical to `mill.modules.Assembly.Rule.Append#separator`

### CODE REVIEW

**Comment by yakivy** [comment]
I can fix `apply` and `this` binary backward compatibility in trade of readability, but `copy` and `unapply` will still fail. Let me know if we need this.

**Comment by lolgab** [comment]
Can you write some motivation for the change you're proposing?

**Comment by yakivy** [comment]
> Can you write some motivation for the change you're proposing?

As mentioned in PR description, motivation is identical to `mill.modules.Assembly.Rule.Append#separator`: some files (in my case service files) doesn't have new line at the end, so they merged into single line.

File: main/src/mill/modules/Assembly.scala [hunk 1]
```diff
@@ -27,9 +27,28 @@ object Assembly {
...
+      def apply(pattern: String, separator: String): AppendPattern =
+        new AppendPattern(Pattern.compile(pattern), separator)
+
+      @deprecated(message = "Binary compatibility shim. To be removed", since = "mill 0.10.1")
+      def unapply(value: AppendPattern): Option[Pattern] = Some(value.pattern)
+    }
+    class AppendPattern(val pattern: Pattern, val separator: String) extends Rule {
+      def this(pattern: Pattern) = this(pattern, defaultSeparator)
+
+      override def productArity: Int = 2
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:41]: This should be deprecated with the same message as below.

File: main/src/mill/modules/Assembly.scala [hunk 2]
```diff
@@ -27,9 +27,28 @@ object Assembly {
...
+      override def productElement(n: Int): Any = n match {
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:42]: This should be deprecated with the same message as below.

File: main/src/mill/modules/Assembly.scala [hunk 3]
```diff
@@ -27,9 +27,28 @@ object Assembly {
...
+        case 0 => pattern
+        case 1 => separator
+        case _ => throw new IndexOutOfBoundsException(n.toString)
+      }
+      override def canEqual(that: Any): Boolean = that.isInstanceOf[AppendPattern]
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:47]: This should be deprecated with the same message as below.

File: main/src/mill/modules/Assembly.scala [hunk 4]
```diff
@@ -27,9 +27,28 @@ object Assembly {
...
+
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:48]: You need to implement also `toString`, `equals`, and `hashCode`

File: main/src/mill/modules/Assembly.scala [hunk 5]
```diff
@@ -50,26 +69,23 @@ object Assembly {
...
+    val matchPatterns = assemblyRules.collect {
+      case r : Rule.AppendPattern => r.pattern.asPredicate() -> r
+      case r @ Rule.ExcludePattern(pattern) => pattern.asPredicate() -> r
     }
 
     mappings.foldLeft(Map.empty[String, GroupedEntry]) {
       case (entries, (mapping, entry)) =>
-        rulesMap.get(mapping) match {
-          case Some(_: Assembly.Rule.Exclude) =>
+        rulesMap.get(mapping) orElse matchPatterns.find(_._1.test(mapping)).map(_._2) match {
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:79]: You could split the two parts into two properly named variables and then do `orElse` separately.

**Review by lolgab — CHANGES_REQUESTED** [review]
There are some missing things around binary compatibility.
I will leave the review of the assembly part to someone else (maybe @lefou ?) since I'm not familiar with that code.
Also, can you add a test that exercises the use case you have (when you need to pass a special separator)?

Thank you very much and sorry for the many requests.

File: main/src/mill/modules/Assembly.scala [hunk 6]
```diff
@@ -27,9 +27,28 @@ object Assembly {
...
-      def apply(pattern: String): AppendPattern = AppendPattern(Pattern.compile(pattern))
+      def apply(pattern: Pattern): AppendPattern = new AppendPattern(pattern, defaultSeparator)
+      def apply(pattern: String): AppendPattern = apply(pattern, defaultSeparator)
+      def apply(pattern: String, separator: String): AppendPattern =
+        new AppendPattern(Pattern.compile(pattern), separator)
+
+      @deprecated(message = "Binary compatibility shim. To be removed", since = "mill 0.10.1")
+      def unapply(value: AppendPattern): Option[Pattern] = Some(value.pattern)
+    }
+    class AppendPattern(val pattern: Pattern, val separator: String) extends Rule {
```
[the remark this answers is not in the document]
Comment (yakivy) [reply main/src/mill/modules/Assembly.scala]: Solution with emulating case class binary compatibility with simple class become dirtier and dirtier, so I created separate `mill.modules.Assembly.Rule.AppendByPattern` class and deprecated `mill.modules.Assembly.Rule.AppendPattern`. Let me know if you don't like this idea.

File: main/src/mill/modules/Assembly.scala (hunk 5 above)
Comment (yakivy) [reply main/src/mill/modules/Assembly.scala:69]: Want to notice that argument in `scala.Option#orElse` is "by-name", so extracting variables will degrade performance, or did you want me to extract functions?

**Comment by yakivy** [comment]
I also found and fixed a bug: `mill.modules.Assembly.Rule.Append` ignores separator if merged files are in different assembly units.

File: main/src/mill/modules/Assembly.scala (hunk 5 above)
Comment (lolgab) [reply main/src/mill/modules/Assembly.scala:69]: Then make the parameter a `def` so the result will be equivalent to a by-name parameter.

File: main/src/mill/modules/Assembly.scala [hunk 7]
```diff
@@ -26,11 +26,19 @@ object Assembly {
...
     case class Append(path: String, separator: String = defaultSeparator) extends Rule
 
+    @deprecated("Use mill.modules.Assembly.Rule.AppendByPattern instead", since = "mill 0.10.1")
     object AppendPattern {
       def apply(pattern: String): AppendPattern = AppendPattern(Pattern.compile(pattern))
     }
+    @deprecated("Use mill.modules.Assembly.Rule.AppendByPattern instead", since = "mill 0.10.1")
     case class AppendPattern(pattern: Pattern) extends Rule
 
+    object AppendByPattern {
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:36]: By creating a new `case class` you're creating a new problem for the next person that will need another parameter.
Then we will have either 3 case classes or a deprecated `case class` and a class with all the ugly workaround you reverted.
I think Mill should not add any more `case class`es to the codebase unless there is a very very strong reason to do so.

Comment (yakivy) [reply main/src/mill/modules/Assembly.scala:36]: For me sounds like YAGNI violation, but you know best. Reverting revert...

File: main/src/mill/modules/Assembly.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (yakivy) [reply main/src/mill/modules/Assembly.scala:38]: Why? Case classes don't have private constructors.

File: main/src/mill/modules/Assembly.scala (hunk 1 above)
Comment (yakivy) [reply main/src/mill/modules/Assembly.scala:42]: Why? It's not a shim, valid and usable method, moreover rule extends product.

File: main/src/mill/modules/Assembly.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (lolgab) [reply main/src/mill/modules/Assembly.scala:38]: This was a `case class` but doesn't need to be forever. The idea is to maintainSince `case class`es as bad for binary compatibility we want to get rid of them.
I'm trying to make the public API as small as possible and having two public APIs to define an object is against this principle (in this case `ApplyPattern(pattern, separator)` and `new ApplyPattern(pattern, separator)`).
We have to maintain the existing ones, but we don't want to create new ones unless it's necessary.

File: main/src/mill/modules/Assembly.scala [hunk 8]
```diff
@@ -27,9 +27,35 @@ object Assembly {
...
+      def apply(pattern: Pattern): AppendPattern = new AppendPattern(pattern, defaultSeparator)
+      def apply(pattern: String): AppendPattern = apply(pattern, defaultSeparator)
+      def apply(pattern: String, separator: String): AppendPattern =
+        new AppendPattern(Pattern.compile(pattern), separator)
+
+      @deprecated(message = "Binary compatibility shim. Don't use it. To be removed", since = "mill 0.10.1")
+      def unapply(value: AppendPattern): Option[Pattern] = Some(value.pattern)
+    }
+    class AppendPattern(val pattern: Pattern, val separator: String) extends Rule {
+      def this(pattern: Pattern) = this(pattern, defaultSeparator)
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:39]: This constructor should be `deprecated` too.

File: main/src/mill/modules/Assembly.scala (hunk 1 above)
Comment (lolgab) [reply main/src/mill/modules/Assembly.scala:42]: > moreover rule extends product.

Then you're right. We can deprecate the product things later for `Rule` and its subclasses.

File: main/src/mill/modules/Assembly.scala [hunk 9]
```diff
@@ -50,26 +76,25 @@ object Assembly {
...
     }.toMap
 
-    val appendPatterns = assemblyRules.collect {
-      case Rule.AppendPattern(pattern) => pattern.asPredicate().test(_)
-    }
-
-    val excludePatterns = assemblyRules.collect {
-      case Rule.ExcludePattern(pattern) => pattern.asPredicate().test(_)
+    val matchPatterns = assemblyRules.collect {
+      case r : Rule.AppendPattern => r.pattern.asPredicate() -> r
```
Comment (lolgab) [inline main/src/mill/modules/Assembly.scala:80]: Can you scalafmt on the changed files?

File: scalalib/test/src/HelloWorldTests.scala [hunk 10]
```diff
@@ -757,6 +765,28 @@ object HelloWorldTests extends TestSuite {
...
+            val Right((result, _)) = eval.apply(target)
+
+            Using.resource(new JarFile(result.path.toIO)) { jarFile =>
+              assert(jarEntries(jarFile).contains("without-new-line.conf"))
+
+              val fileContent = readFileFromJar(jarFile, "without-new-line.conf")
+
+              assert(
+                "without-new-line\\.first=first(?:\n|$)".r.findFirstIn(fileContent).isDefined,
+                "without-new-line\\.second=second(?:\n|$)".r.findFirstIn(fileContent).isDefined,
```
Comment (lolgab) [inline scalalib/test/src/HelloWorldTests.scala:785]: Can you assert the exact content of the file instead?
Like:
```scala
val result  = readFileFromJar(jarFile, "without-new-line.conf")
val expected = "without-new-line.first=first\nwithout-new-line.second=second"
assert(result == expected)

```

**Review by lolgab — CHANGES_REQUESTED** [review]
Some minor comments, otherwise it looks good to me.

File: scalalib/test/src/HelloWorldTests.scala (hunk 10 above)
Comment (lolgab) [reply scalalib/test/src/HelloWorldTests.scala:785]: Gotcha! Then what do you think about:
```scala
val result  = readFileFromJar(jarFile, "without-new-line.conf").split('\n').toSet
val expected = Set("without-new-line.first=first", "without-new-line.second=second")
assert(result == expected)
```
?

**Review by lolgab — APPROVED** [review]
LGTM! 👍
Thank you for your contribution @yakivy

**Review by lefou — APPROVED** [review]
I see the efforts you already put into this, so I guess it's good to merge. Thank you!

Besides, I don't like the bloat we add with this kind of patches. It's a maintainers nightmare. And will make future contributions much harder. We might evaluate some other approaches for powerful case class-like config objects which allow more binary compatibility, e.g. data-class or something like this.

**Comment by lolgab** [comment]
@lefou The idea is to delete all the bloat of `case class` methods before `0.11` and just have a class with manual `apply`, `toString` and few more methods. It's the only way I know to evolve a codebase like the mill one. The Sbt team uses code-generation but it has its own tradeoffs.

**Pull request merged**